File: occasum-core/src/main/java/pw/stamina/occasum/properties/collection/SetProperty.java
package pw.stamina.occasum.properties.collection;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public final class SetProperty<T> extends AbstractCollectionProperty<T, Set<T>> {

    SetProperty(String name, Set<T> elements, CollectionPropertyAdapter<T> adapter) {
        super(name, elements, adapter);
    }

    @Override
    public Set<T> getElements() {
        return Collections.unmodifiableSet(elements);
    }

    public static <T> SetProperty<T> hash(String name, CollectionPropertyAdapter<T> adapter) {
        return new SetProperty<>(name, new HashSet<>(), adapter);
    }

    public static <T> SetProperty<T> concurrentHash(String name, CollectionPropertyAdapter<T> adapter) {
        return new SetProperty<>(name, ConcurrentHashMap.newKeySet(), adapter);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/collection/SetProperty.kt
package pw.stamina.occasum.properties.collection

import java.util.*
import java.util.concurrent.ConcurrentHashMap

class SetProperty<T> internal constructor(
        name: String,
        elements: Set<T>,
        adapter: CollectionPropertyAdapter<T>)
    : AbstractCollectionProperty<T, Set<T>>(name, elements, adapter) {

    override val elements = elements
        get() = Collections.unmodifiableSet(field)

    companion object {

        fun <T> hash(name: String, adapter: CollectionPropertyAdapter<T>): SetProperty<T> {
            return SetProperty(name, HashSet(), adapter)
        }

        fun <T> concurrentHash(name: String, adapter: CollectionPropertyAdapter<T>): SetProperty<T> {
            return SetProperty(name, ConcurrentHashMap.newKeySet(), adapter)
        }
    }
}
